### PR TITLE
Amda dev

### PR DESCRIPTION
**Pull request by Dolgalad**
added functionality to speasy.amda module : 
 - User private parameters
 - Product types : TimeTable, Catalog, Parameter, Dataset
 - Query functions : list_parameters, list_datasets, list_timetables, list_catalogs, list_user_timetables, list_user_catalogs, list_user_parameters
 - Getter functions: get_parameter, get_dataset, get_timetable, get_catalog, get_user_timetable, get_user_parameter, get_user_catalog
 - ObsDataTreeParser replaced by InventoryTree class : used for parsing timetable and catalog XML
 - Documentation (mainly speasy.amda, speasy.common.variable)
 - examples and tests

### CODE REVIEW

File: speasy/config/__init__.py [hunk 1]
```diff
@@ -1,24 +1,50 @@
...
+        :rtype: str
+        """
         if self.key1 in _config and self.key2 in _config[self.key1]:
             return _config[self.key1][self.key2]
         else:
-            return self.default
+            if len(self.default):
+                return self.default
+            # key is not defined, raise an UndefinedConfigEntry exception
+            raise UndefinedConfigEntry(key1=self.key1, key2=self.key2, default=self.default)
```
Comment (jeandet) [inline speasy/config/__init__.py:47]: This should not raise here, most of the time this is ok to have empty config entry.

File: speasy/config/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
+spease.config.exceptions
+------------------------
+
+Exception definitions.
+"""
+
+class UndefinedConfigEntry(Exception):
```
Comment (jeandet) [inline speasy/config/exceptions.py:8]: I don't think this is useful

File: tests/test_amda_parameter.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        self.ws = AMDA()
+        self.start=datetime(2000,1,1,1,1)
+        self.stop=datetime(2000,1,1,1,2)
+        self.data = self.ws.get_parameter("imf", self.start, self.stop)
+        self.dataset = self.ws.get_dataset("ace-imf-all", self.start, self.stop)
+
+
+    def tearDown(self):
+        pass
+    # check that the data is not None
```
Comment (jeandet) [inline tests/test_amda_parameter.py:29]: You don't need to comment if it doesn't add information that the code doesn't provide.

File: examples/amda_list_catalogs.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+from speasy.amda import AMDA
+import datetime
+
+amda = AMDA()
+ttlist = amda.list_catalogs()
+for catalog in amda.list_catalogs():
+    print(catalog)
+    print(amda.catalog[catalog])
+    #print(amda.get_catalog(catalog).data)
```
Comment (jeandet) [inline examples/amda_list_catalogs.py:12]:
```suggestion
```

**Review by jeandet — COMMENTED** [review]
Good first PR! :) 
The config entry exception should be removed then most comments can either be resolved in this PR or fixed later.

File: speasy/amda/rest.py [hunk 5]
```diff
@@ -1,37 +1,245 @@
...
+        :type n_try: int
+        :return: request result, stripped of spaces and newlines
+        :rtype: str
+        """
+        url+="token={}".format(self.get_token)
+        for _ in [None]*n_try: # in case of failure
+            # add token now ? does it change
+            log.debug(f"Send request on AMDA server {url}")
+            r = requests.get(url)
+            r = requests.get(r.text.strip())
```
[the remark this answers is not in the document]
Comment (Dolgalad) [reply speasy/amda/rest.py:112]: I should ask Benjamin about this, seems sloppy and could be changed in the near future

File: speasy/config/__init__.py (hunk 1 above)
Comment (Dolgalad) [reply speasy/config/__init__.py]: ok. I have reverted the code to its older version.

**Pull request merged**